### PR TITLE
Recompute flat_arg_fake_tensors after fakification

### PR DESCRIPTION
**Pull request by andrewor14**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98769

Summary: This fixes the case when some of the input tensors were
real tensors and fakified in `validate_and_convert_non_fake_tensors`,
but `flat_arg_fake_tensors` would not contain all the inputs
because it was computed before the fakification. We fix this by
recomputing `flat_arg_fake_tensors` after fakification as well.

Test Plan:
python test/dynamo/test_export.py ExportTests.test_mixed_real_and_fake_inputs
    
Reviewers: Chillee, voznesenskym

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -1161,6 +1161,10 @@ def dispatch(self, func, types, args=(), kwargs=None):
             func, converter, args, kwargs
         )
 
+        # Recompute flat_arg_fake_tensors here again in case some of the inputs
+        # were real tensors and fakified in validate_and_convert_non_fake_tensors
+        flat_arg_fake_tensors = tree_flatten_only(FakeTensor, (args, kwargs))
+
```
Comment (voznesenskym) [inline torch/_subclasses/fake_tensor.py:1167]: This is not free! What is the perf overhead here? Do you need help kicking off a measurement run?

Comment (andrewor14) [reply torch/_subclasses/fake_tensor.py:1167]: Yes, that would be great. Thanks.

Comment (eellison) [reply torch/_subclasses/fake_tensor.py:1167]: You could move the creation of this to `validate_and_convert_non_fake_tensors` ? would be faster ?

**Review by eellison — COMMENTED** [review]
lgtm but per @voznesenskym's concerns we should be careful about adding more pytree calls since they are slow

**Review by eellison — COMMENTED** [review]
Note that this will prevent real_tensors from being used with constant ops, but I think that's fine

**Comment by voznesenskym** [comment]
Perf run here https://github.com/pytorch/pytorch/actions/runs/4691840608

File: test/dynamo/test_export.py [hunk 2]
```diff
@@ -2389,6 +2389,36 @@ def f_correct(x):
...
+                scaled_weight = self.conv.weight * scale_factor.reshape(weight_shape)
+                zero_bias = torch.zeros_like(self.conv.bias, dtype=input.dtype)
+                conv = self.conv._conv_forward(input, scaled_weight, zero_bias)
+                conv_orig = conv / scale_factor.reshape(bias_shape)
+                conv_orig = conv_orig + self.conv.bias.reshape(bias_shape)
+                conv = self.bn(conv_orig)
+                return conv
+
+        example_inputs = (torch.randn(1, 1, 3, 3),)
+        torch._dynamo.export(
```
[the remark this answers is not in the document]
Comment (andrewor14) [reply test/dynamo/test_export.py:2492]: Can we do this in a separate PR? This is currently blocking https://github.com/pytorch/pytorch/pull/98568 right now

File: torch/_subclasses/fake_tensor.py [hunk 3]
```diff
@@ -1157,7 +1157,7 @@ def dispatch(self, func, types, args=(), kwargs=None):
 
             return converter(self, args[0])
 
-        args, kwargs = self.validate_and_convert_non_fake_tensors(
+        args, kwargs, flat_arg_fake_tensors = self.validate_and_convert_non_fake_tensors(
```
Comment (eellison) [inline torch/_subclasses/fake_tensor.py:1160]: Can you add back the comment you had earlier ?

**Review by voznesenskym — APPROVED** [review]

File: test/dynamo/test_export.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (voznesenskym) [reply test/dynamo/test_export.py:2492]: Followup is fine by me, I trust you to do it :)

[the remark this answers is not in the document]
Comment (eellison) [reply test/dynamo/test_export.py:2492]: The fact that there wasn't a good fake tensor only test on the original PR https://github.com/pytorch/pytorch/pull/90417 is probably we have this error lol. Can you please add it in this PR ?

**Review by eellison — COMMENTED** [review]
looks good but please add separate test in test_fake_tensor

File: torch/_subclasses/fake_tensor.py (hunk 3 above)
Comment (andrewor14) [reply torch/_subclasses/fake_tensor.py:1162]: done

**Comment by andrewor14** [comment]
Ok, I'm merging this. Thanks everyone for your feedback!

**Pull request closed without merging**